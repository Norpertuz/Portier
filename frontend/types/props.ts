import { PropsWithChildren } from 'react';

import { Language } from './common';

export interface BaseProps {
  className?: string;
}

export interface ButtonProps extends BaseProps, NavigationItemProps {
  children: string;
  size?: 'md' | 'sm';
}

export type ContainerProps = BaseProps & PropsWithChildren;

export interface HamburgerProps {
  isOpened: boolean;
  onClick: () => void;
}

export type LayoutProps = Language & PropsWithChildren;

export interface NavigationItemProps
  extends PropsWithChildren {
  url: string;
}
