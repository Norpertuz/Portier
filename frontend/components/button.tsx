import Link from 'next/link';
import { FC } from 'react';

import { classNames } from '#data/functions';
import styles from '#styles/common.module.scss';
import { ButtonProps } from '#types/props';

const Button: FC<ButtonProps> = ({
  children, className = '', size = 'md', url
}) => (
  <Link
    className={classNames([size === 'md' ? styles.button : styles['button--sm'], className])}
    data-testid="button"
    href={url}
  >
    { children }
  </Link>
);

export { Button };
