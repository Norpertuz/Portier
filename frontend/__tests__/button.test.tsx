import { render, waitFor } from '@testing-library/react';
import renderer from 'react-test-renderer';

import { Button } from '#components/button';
import { ButtonProps } from '#types/props';

type TestData = {
  [variant in 'initialization' | 'medium' | 'small']: ButtonProps;
};

const testData: TestData = {
  initialization: { children: 'Initialize', url: '/portfolio' },
  medium: { children: 'Medium Button', size: 'md', url: '/' },
  small: { children: 'Small Button', size: 'sm', url: '/about-me' }
};

describe('Button Component', () => {
  it('initialize a component with default props', async () => {
    const { initialization } = testData;
    const { getByTestId } = render(<Button {...initialization} />);
    const defaultButton = getByTestId('button');

    const tree = renderer.create(<Button {...initialization} />).toJSON();

    await waitFor(() => {
      expect(defaultButton).toBeInTheDocument();
      expect(defaultButton).toHaveTextContent(initialization.children);
      expect(defaultButton).toHaveAttribute('class', 'button');
      expect(defaultButton).toHaveAttribute('href', initialization.url);
      expect(tree).toMatchSnapshot();
    });
  });

  it('renders the standard button', async () => {
    const { medium } = testData;
    const { getByTestId } = render(<Button {...medium} />);
    const mediumButton = getByTestId('button');

    const tree = renderer.create(<Button {...medium} />).toJSON();

    await waitFor(() => {
      expect(mediumButton).toBeInTheDocument();
      expect(mediumButton).toHaveAttribute('class', 'button');
      expect(tree).toMatchSnapshot();
    });
  });

  it('renders the smaller button', async () => {
    const { small } = testData;
    const { getByTestId } = render(<Button {...small} />);
    const smallButton = getByTestId('button');

    const tree = renderer.create(<Button {...small} />).toJSON();

    await waitFor(() => {
      expect(smallButton).toBeInTheDocument();
      expect(smallButton).toHaveAttribute('class', 'button--sm');
      expect(tree).toMatchSnapshot();
    });
  });
});
