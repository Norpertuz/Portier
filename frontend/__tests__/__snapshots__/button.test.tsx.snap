// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component initialize a component with default props 1`] = `
<a
  className="button"
  data-testid="button"
  href="/portfolio"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onTouchStart={[Function]}
>
  Initialize
</a>
`;

exports[`Button Component renders the smaller button 1`] = `
<a
  className="button--sm"
  data-testid="button"
  href="/about-me"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onTouchStart={[Function]}
>
  Small Button
</a>
`;

exports[`Button Component renders the standard button 1`] = `
<a
  className="button"
  data-testid="button"
  href="/"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onTouchStart={[Function]}
>
  Medium Button
</a>
`;
